File: src/contrib/hedwig/client/src/main/cpp/test/utiltest.cpp
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <cppunit/Test.h>
#include <cppunit/TestSuite.h>
#include <cppunit/extensions/HelperMacros.h>

#include "../lib/util.h"
#include <hedwig/exceptions.h>
#include <stdexcept>

using namespace CppUnit;

class UtilTestSuite : public CppUnit::TestFixture {
  CPPUNIT_TEST_SUITE( UtilTestSuite );
  CPPUNIT_TEST(testHostAddress);
  CPPUNIT_TEST_SUITE_END();

public:
  void setUp()
  {
  }

  void tearDown() 
  {
  }

  void testHostAddress() {
    // good address (no ports)
    Hedwig::HostAddress a1 = Hedwig::HostAddress::fromString("www.yahoo.com");
    CPPUNIT_ASSERT(a1.port() == 4080);

    // good address with ip (no ports)
    Hedwig::HostAddress a2 = Hedwig::HostAddress::fromString("127.0.0.1");
    CPPUNIT_ASSERT(a2.port() == 4080);
    CPPUNIT_ASSERT(a2.ip() == ((127 << 24) | 1));

    // good address
    Hedwig::HostAddress a3 = Hedwig::HostAddress::fromString("www.yahoo.com:80");
    CPPUNIT_ASSERT(a3.port() == 80);

    // good address with ip
    Hedwig::HostAddress a4 = Hedwig::HostAddress::fromString("127.0.0.1:80");
    CPPUNIT_ASSERT(a4.port() == 80);
    CPPUNIT_ASSERT(a4.ip() == ((127 << 24) | 1));

    // good address (with ssl)
    Hedwig::HostAddress a5 = Hedwig::HostAddress::fromString("www.yahoo.com:80:443");
    CPPUNIT_ASSERT(a5.port() == 80);

    // good address with ip
    Hedwig::HostAddress a6 = Hedwig::HostAddress::fromString("127.0.0.1:80:443");
    CPPUNIT_ASSERT(a6.port() == 80);
    CPPUNIT_ASSERT(a6.ip() == ((127 << 24) | 1));

    // nothing
    CPPUNIT_ASSERT_THROW(Hedwig::HostAddress::fromString(""), Hedwig::HostResolutionException);
    
    // nothing but colons
    CPPUNIT_ASSERT_THROW(Hedwig::HostAddress::fromString("::::::::::::::::"), Hedwig::ConfigurationException);
    
    // only port number
    CPPUNIT_ASSERT_THROW(Hedwig::HostAddress::fromString(":80"), Hedwig::HostResolutionException);
 
    // text after colon (isn't supported)
    CPPUNIT_ASSERT_THROW(Hedwig::HostAddress::fromString("www.yahoo.com:http"), Hedwig::ConfigurationException);
    
    // invalid hostname
    CPPUNIT_ASSERT_THROW(Hedwig::HostAddress::fromString("com.oohay.www:80"), Hedwig::HostResolutionException);
    
    // null
    CPPUNIT_ASSERT_THROW(Hedwig::HostAddress::fromString(NULL), std::logic_error);
  }
};

CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( UtilTestSuite, "Util" );

File: src/java/test/org/apache/zookeeper/server/persistence/UtilTest.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.zookeeper.server.persistence;

import org.apache.zookeeper.test.ClientBase;
import org.junit.Assert;
import org.junit.Test;

import java.io.File;
import java.io.IOException;
import java.util.List;

public class UtilTest {
    @Test
    public void testSortDataDirAscending() {
        File[] files = new File[5];

        files[0] = new File("foo.10027c6de");
        files[1] = new File("foo.10027c6df");
        files[2] = new File("bar.10027c6dd");
        files[3] = new File("foo.10027c6dc");
        files[4] = new File("foo.20027c6dc");

        File[] orig = files.clone();

        List<File> filelist = Util.sortDataDir(files, "foo", true);

        Assert.assertEquals(orig[2], filelist.get(0));
        Assert.assertEquals(orig[3], filelist.get(1));
        Assert.assertEquals(orig[0], filelist.get(2));
        Assert.assertEquals(orig[1], filelist.get(3));
        Assert.assertEquals(orig[4], filelist.get(4));
    }

    @Test
    public void testSortDataDirDescending() {
        File[] files = new File[5];

        files[0] = new File("foo.10027c6de");
        files[1] = new File("foo.10027c6df");
        files[2] = new File("bar.10027c6dd");
        files[3] = new File("foo.10027c6dc");
        files[4] = new File("foo.20027c6dc");

        File[] orig = files.clone();

        List<File> filelist = Util.sortDataDir(files, "foo", false);

        Assert.assertEquals(orig[4], filelist.get(0));
        Assert.assertEquals(orig[1], filelist.get(1));
        Assert.assertEquals(orig[0], filelist.get(2));
        Assert.assertEquals(orig[3], filelist.get(3));
        Assert.assertEquals(orig[2], filelist.get(4));
    }

    @Test
    public void testInvalidSnapshot() {
        File f = null;
        File tmpFileDir = null;
        try {
            tmpFileDir = ClientBase.createTmpDir();
            f = new File(tmpFileDir, "snapshot.0");
            if (!f.exists()) {
                f.createNewFile();
            }
            Assert.assertFalse("Snapshot file size is greater than 9 bytes", Util.isValidSnapshot(f));
            Assert.assertTrue("Can't delete file", f.delete());
        } catch (IOException e) {
        } finally {
            if (null != tmpFileDir) {
                ClientBase.recursiveDelete(tmpFileDir);
            }
        }
    }
}
